src/main/java/com/google/cloud/genomics/dockerflow/task/Task.java

```diff
@@ -390,6 +388,151 @@ private void resolveLocalCopies() {
+
+      // Folder copying is done by editing the Docker command, 
+      // not by automated file staging
+      if (val != null && p.isFolder()) {
+        gcsPaths.put(p.getName(), val);
+        localPaths.put(p.getName(), p.getLocalCopy().getPath());
+        
+        // Replace var in user script with local path
+        defn.getDocker().setCmd(
+            defn.getDocker().getCmd().replace(
+                "${" + p.getName() + "}", 
+                DockerflowConstants.DEFAULT_MOUNT_POINT + 
+                p.getLocalCopy().getPath() + "\""));
+
+        // Turn it into an env var
+        p.setLocalCopy(null);
+        p.setType(null);
+                
+        // Create input environment variables for each output folder
+        // Remove the folder option, since it's used by Dockerflow,
+        // and is not recognized by Pipelines API
+        if (!inputs.contains(p.getName())) {
+          defn.getOutputParameters().remove(p);
+          if (defn.getInputParameters() == null) {
+            defn.setInputParameters(new LinkedHashSet<Param>());
+          }
+          defn.getInputParameters().add(p);
+          
+          // Move output folder args to input args
+          if (args.getOutputs() != null && args.getOutputs().containsKey(p.getName())) {
+            if (args.getInputs() == null) {
+              args.setInputs(new LinkedHashMap<String,String>());
+            }
+            args.getInputs().put(p.getName(), args.getOutputs().get(p.getName()));
+            args.getOutputs().remove(p.getName());
+          }
+        }
+      }
+    }
+    
+    // Pipelines API rejects an empty parameter set
+    if (defn.getOutputParameters() != null && defn.getOutputParameters().isEmpty()) {
+      defn.setOutputParameters(null);
+    }
+
+    final String INSTALL_GSUTIL = 
+        "# Install gsutil\n" +
+        "if ! type gsutil; then\n" +
+        "  apt-get update\n" +
+        "  apt-get --yes install apt-utils gcc python-dev python-setuptools ca-certificates\n" +
+        "  easy_install -U pip\n" +
+        "  pip install -U crcmod\n" +
+        "\n" +
+        "  apt-get --yes install lsb-release\n" +
+        "  export CLOUD_SDK_REPO=\"cloud-sdk-$(lsb_release -c -s)\"\n" +
+        "  echo \"deb http://packages.cloud.google.com/apt $CLOUD_SDK_REPO main\" >> /etc/apt/sources.list.d/google-cloud-sdk.list\n" +
+        "  apt-get update && apt-get --yes install curl\n" +
+        "  curl https://packages.cloud.google.com/apt/doc/apt-key.gpg | apt-key add -\n" +
+        "  apt-get update && apt-get --yes install google-cloud-sdk\n" +
+        "fi\n";
+
+    // Edit the Docker command to copy the folders
+    StringBuilder copyInputs = new StringBuilder();
+    StringBuilder copyOutputs = new StringBuilder();
+    
+    for (String name : gcsPaths.keySet()) {
+      // Copy input folder
+      if (inputs.contains(name)) {
+        copyInputs.append(
+            "mkdir -p " +
```

Comment:
Sorry - I had one more comment that seems to have gotten lost.

I suggest reformatting this as:

    copyInputs.append(
        String.format(
            "mkdir -p %s/%s\n",
            DockerflowConstants.DEFAULT_MOUNT_POINT,
            localPaths.get(name)) +
        String.format(
            "for ((i = 0; i < 3; i++)); do\n" +
            "  if gsutil -m rsync -r %s/ %s/%s; then\n" +
            "    break\n" +
            "  elif ((i == 2)); then\n" +
            "    2>&1 echo \"Recursive localization failed.\"\n" +
            "    exit 1\n" +
            "  fi\n",
            gcsPaths.get(name),
            DockerflowConstants.DEFAULT_MOUNT_POINT,
            localPaths.get(name)))

to make it more clear what the generated shell script will look like.